File: src/org/perses/reduction/ReductionConfiguration.java
/*
 * Copyright (C) 2003-2017 Chengnian Sun.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.reduction;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Joiner;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.perses.grammar.AbstractParserFacade;
import org.perses.grammar.ParserFacadeFactory;
import org.perses.program.SourceFile;
import org.perses.util.Util;

import java.io.File;
import java.lang.reflect.Field;
import java.time.LocalDateTime;
import java.util.Optional;

import static com.google.common.base.Preconditions.*;

/**
 * This is the internal configuration for Perses reducer.
 *
 * <p>TODO: refactor this to a AutoValue class, or use the BUILDER pattern.
 */
public final class ReductionConfiguration {

  private static final String PERSES_TEMP_ROOT_PREFIX = "PersesTempRoot";
  /** The working directory of the reduction. */
  private final File workingFolder;
  /** The test script for reduction */
  private final SourceFile testScript;
  /** The file to reduce */
  private final SourceFile fileToReduce;
  /** The parser facade. */
  private final AbstractParserFacade parserFacade;

  private final boolean keepOriginalCodeFormat;
  private final boolean fixpointReduction;
  private final boolean enableTestScriptExecutionCaching;
  private final boolean useRealDeltaDebugger;
  private final boolean refreshQueryCache;
  private final int maxReductionLevel;
  /** This file is used to store the best result of reduction. */
  private final File bestResultFile;

  private final File tempRootFolder;
  @Nullable private final File statisticsFile;
  @Nullable private final File progressDumpFile;
  private final int numberOfThreads;

  private final IPartitionReductionPolicy.MultiNodePartitionReductionPolicy
      multiNodePartitionReductionPolicy;
  private final IPartitionReductionPolicy.SingleNodePartitionReductionPolicy
      singleNodePartitionReductionPolicy;

  public ReductionConfiguration(
      final File workingFolder,
      final File testScriptFile,
      final File fileToReduce,
      final File bestResultFile,
      @Nullable final File statisticsFile,
      @Nullable final File progressDumpFile,
      final boolean keepOriginalCodeFormat,
      final boolean fixpointReduction,
      final boolean enableTestScriptExecutionCaching,
      final boolean useRealDeltaDebugger,
      final boolean refreshQueryCache,
      final int maxReductionLevel,
      final int numOfReductionThreads,
      IPartitionReductionPolicy.MultiNodePartitionReductionPolicy multiNodePartitionReductionPolicy,
      IPartitionReductionPolicy.SingleNodePartitionReductionPolicy
          singleNodePartitionReductionPolicy) {
    this.workingFolder = workingFolder;
    this.testScript = SourceFile.createFromPath(testScriptFile);
    this.fileToReduce = SourceFile.createFromPath(fileToReduce);
    this.parserFacade =
        ParserFacadeFactory.SINGLETON.createParserFacade(this.fileToReduce.getLanguageKind());

    this.bestResultFile = bestResultFile;
    this.statisticsFile = statisticsFile;

    this.keepOriginalCodeFormat = keepOriginalCodeFormat;
    this.fixpointReduction = fixpointReduction;
    this.enableTestScriptExecutionCaching = enableTestScriptExecutionCaching;
    this.useRealDeltaDebugger = useRealDeltaDebugger;
    this.refreshQueryCache = refreshQueryCache;

    this.maxReductionLevel = maxReductionLevel;
    this.multiNodePartitionReductionPolicy = checkNotNull(multiNodePartitionReductionPolicy);
    this.singleNodePartitionReductionPolicy = checkNotNull(singleNodePartitionReductionPolicy);
    checkArgument(
        this.workingFolder.isDirectory(),
        "The working folder is not a directory: %s.",
        this.workingFolder);
    checkArgument(numOfReductionThreads > 0, "The number of reduction threads must be positive.");
    this.numberOfThreads = numOfReductionThreads;
    tempRootFolder =
        new File(
            workingFolder,
            getTempRootFolderName(
                fileToReduce.getName(), testScriptFile.getName(), LocalDateTime.now()));
    this.progressDumpFile = progressDumpFile;
    validateConfiguration();
  }

  public void validateConfiguration() {
    checkState(workingFolder.exists(), "The working directory %s does not exist.", workingFolder);
    checkState(
        workingFolder.isDirectory(), "The working directory %s is not a directory", workingFolder);
    checkState(
        maxReductionLevel > 0, "The max reduction level should be positive: %s", maxReductionLevel);
    checkNotNull(multiNodePartitionReductionPolicy);
    checkNotNull(singleNodePartitionReductionPolicy);
  }

  public boolean isKeepOriginalCodeFormat() {
    return keepOriginalCodeFormat;
  }

  public IPartitionReductionPolicy.MultiNodePartitionReductionPolicy
      getMultiNodePartitionReductionPolicy() {
    return multiNodePartitionReductionPolicy;
  }

  public IPartitionReductionPolicy.SingleNodePartitionReductionPolicy
      getSingleNodePartitionReductionPolicy() {
    return singleNodePartitionReductionPolicy;
  }

  public int getNumOfReductionThreads() {
    return this.numberOfThreads;
  }

  public File getTempRootFolder() {
    return tempRootFolder;
  }

  public Optional<File> getStatisticsFile() {
    return Optional.ofNullable(statisticsFile);
  }

  public Optional<File> getProgressDumpFile() {
    return Optional.ofNullable(progressDumpFile);
  }

  public Optional<File> getTestScriptStatisticsFile() {
    if (statisticsFile == null) {
      return Optional.empty();
    }
    return Optional.of(
        new File(statisticsFile.getParentFile(), "testscript-" + statisticsFile.getName()));
  }

  public File getWorkingFolder() {
    return workingFolder;
  }

  public SourceFile getTestScript() {
    return testScript;
  }

  public SourceFile getFileToReduce() {
    return fileToReduce;
  }

  public boolean isFixpointReduction() {
    return fixpointReduction;
  }

  public int getMaxReductionLevel() {
    return maxReductionLevel;
  }

  public boolean isUseRealDeltaDebugger() {
    return useRealDeltaDebugger;
  }

  public File getFixpointIterationResultFile(int fixpointIteration) {
    final File bestResultFile = getBestResultFile();
    return new File(
        getTempRootFolder(), bestResultFile.getName() + "_fixpoint_iteration_" + fixpointIteration);
  }

  public File getBestResultFile() {
    return bestResultFile;
  }

  public boolean isRefreshQueryCache() {
    return refreshQueryCache;
  }

  public String dumpConfiguration() {
    final StringBuilder builder = new StringBuilder();
    try {
      for (Field field : ReductionConfiguration.class.getDeclaredFields()) {
        builder.append(field.getName()).append('=').append(field.get(this)).append('\n');
      }
    } catch (IllegalAccessException e) {
      throw new AssertionError(e);
    }
    return builder.toString();
  }

  public boolean isEnableTestScriptExecutionCaching() {
    return enableTestScriptExecutionCaching;
  }

  public AbstractParserFacade getParserFacade() {
    return parserFacade;
  }

  @VisibleForTesting
  static String getTempRootFolderName(
      String fileNameForReduction, String testScriptName, LocalDateTime time) {
    return Joiner.on('_')
        .join(
            PERSES_TEMP_ROOT_PREFIX,
            fileNameForReduction,
            testScriptName,
            Util.formatDateForFileName(time));
  }
}

File: src/org/perses/reduction/ReductionConfiguration.kt
/*
 * Copyright (C) 2018-2026 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.reduction

import org.perses.grammar.AbstractParserFacade
import org.perses.listminimizer.EnumListMinimizerType
import org.perses.program.printer.PrinterRegistry
import org.perses.reduction.reducer.EnumMimirReductionAlgorithm
import org.perses.reduction.semantics.ISemanticsProviderCreator
import org.perses.util.hashing.EnumShaAlgorithm
import java.nio.file.Files
import java.nio.file.Path

/**
 * This is the internal configuration for Perses reducer.
 */
class ReductionConfiguration(
  val globalFixpoint: Boolean,
  val mainReducerAnnotation: AbstractReducerNameAndDesc,
  val cleanupReducerAnnotation: AbstractReducerNameAndDesc?,
  val fixpointReductionForMainReducer: Boolean,
  val enableDeprecatedQueryCaching: Boolean,
  val fullyDeterministicMode: Boolean,
  val numOfReductionThreads: Int,
  val canonicalParserFacade: AbstractParserFacade,
  val persesConfig: PersesConfig,
  val listMinimizerConfig: ListMinimizerConfig,
  val vulcanConfig: VulcanConfig,
  val lprConfig: LPRConfig,
  val latraConfig: LatraConfig,
  val sfcConfig: SfcConfig,
  val levelBasedReducerConfig: LevelBasedReducerConfig,
  val mimirConfig: MimirConfig,
  val shaHashAlgorithm: EnumShaAlgorithm,
) {
  val originalFormatPrinter =
    PrinterRegistry.getPrinter(
      format = canonicalParserFacade.language.origCodeFormatControl,
      lexerAtnWrapper = canonicalParserFacade.lexerAtnWrapper,
    )

  init {
    require(numOfReductionThreads > 0) {
      "The number of reduction threads should be positive: $numOfReductionThreads"
    }
  }

  // Need to generate hashcode and equals.
  data class PersesConfig(
    val enableTopDownReduction: Boolean,
    val enableReducingRegularRuleNode: Boolean,
    val enableReducingKleeneOptionalNode: Boolean,
    val maxEditCountForRegularRuleNode: Int,
    val maxBfsDepthForRegularRuleNode: Int,
    val stopAtFirstCompatibleChildren: Boolean,
    val enableLiteralReplacementForListMinimizer: Boolean,
    val enableLiteralReplacementForRegularRuleNode: Boolean,
    val listMinimizerTypeForKleene: EnumListMinimizerType,
    val anticipatedFinalTokenCount: Int,
  ) {
    init {
      require(maxEditCountForRegularRuleNode > 0)
      require(maxBfsDepthForRegularRuleNode > 0)
      require(anticipatedFinalTokenCount >= 0) { "Anticipated token count must be at least 0" }
    }
  }

  // Need to generate hashcode and equals.
  data class ListMinimizerConfig(
    val minSlidingWindowSize: Int,
    val maxSlidingWindowSize: Int,
  ) {
    init {
      require(0 < minSlidingWindowSize) { minSlidingWindowSize }
      require(minSlidingWindowSize <= maxSlidingWindowSize) {
        "$minSlidingWindowSize $maxSlidingWindowSize"
      }
    }
  }

  // Need to generate hashcode and equals.
  data class LevelBasedReducerConfig(
    val defaultListMinimizerType: EnumListMinimizerType,
  )

  // Need to generate hashcode and equals.
  data class LatraConfig(
    val listMinimizerForTransformations: EnumListMinimizerType,
  )

  // Need to generate hashcode and equals.
  data class SfcConfig(
    /**
     * A subtree of more tokens than this is left alone: the number of structures a subtree can
     * be converted into grows with the number of subtrees under it that can be reused.
     */
    val subtreeTokenCountLimit: Int,
    /** How many of a subtree's candidates are tested before moving on to the next subtree. */
    val candidateLimit: Int,
  )

  // Need to generate hashcode and equals.
  class MimirConfig(
    val semanticsProviderCreator: ISemanticsProviderCreator?,
    val nodeCountThresholdToUseOneByOne: Int,
    val enableBottomUpReductionAfterMainReductionLoop: Boolean,
    val bottomUpReductionMaxDepth: Int,
    val mimirReductionAlgorithm: EnumMimirReductionAlgorithm,
    val enableMimirForRegularRuleNodes: Boolean,
    /*
     * Delete a definition node, with all its concrete uses.
     * By doing this, we can guarantee we do not introduce use-without-def errors.
     */
    val deleteDefWithAllItsConcreteUses: Boolean,
  ) {
    init {
      require(nodeCountThresholdToUseOneByOne >= 0) {
        "The nodeCount threshold must be greater than or equal to 0"
      }
    }
  }

  // Need to generate hashcode and equals.
  class VulcanConfig(
    val nonDeletionIterationLimit: Int,
    val windowSizeForLocalExhaustivePatternReduction: Int,
    val vulcanFixpoint: Boolean,
  ) {
    init {
      require(nonDeletionIterationLimit > 0) {
        "Non deletion iteration limit: $nonDeletionIterationLimit"
      }
      require(windowSizeForLocalExhaustivePatternReduction > 0) {
        "$windowSizeForLocalExhaustivePatternReduction"
      }
    }
  }

  // Need to generate hashcode and equals.
  class LPRConfig(
    val llmClientPath: Path?,
    val lprFixpoint: Boolean,
  ) {
    init {
      require(llmClientPath == null || Files.isRegularFile(llmClientPath)) {
        "The script to invoke LLM is not a regular file: $llmClientPath"
      }
    }
  }
}
